packages/schemas/src/generate-task-schemas.ts

```diff
@@ -1,24 +1,66 @@
 #!/usr/bin/env node
 
-import fs from 'fs';
+import fs, { FileHandle } from 'fs/promises';
 import path from 'path';
+import { templateJsonSchemaWithFiles } from './generate-schemas';
+import generateTypes from './generate-types';
 
-const {
-  templateJsonSchemaWithFiles,
-} = require('./generate-schemas');
+const queueGranulesFilesJsonSchema = require('../preSyncedFiles.schema.json');
 
-const taskDirectory = process.argv[2];
-const command = process.argv[3];
-const taskSchemasDirectory = path.join(taskDirectory, 'schemas');
+const TYPESCRIPT_OPTION = '--typescript';
+function parseOptions(taskDirectory: string, args: string[]): { tsFile?: Promise<FileHandle> } {
+  let tsFile;
 
-if (command === 'files') {
-  const schemaTemplateFiles = fs.readdirSync(taskSchemasDirectory)
-    .filter((filename) => filename.endsWith('.template'));
+  if (args.includes(TYPESCRIPT_OPTION)) {
+    const position = args.indexOf(TYPESCRIPT_OPTION);
+    if (args.length <= position + 1) {
+      throw new Error('Missing filepath for typescript option');
+    }
+    tsFile = fs.open(path.join(taskDirectory, args[position + 1]), 'w');
+  }
+  return { tsFile };
+}
 
-  schemaTemplateFiles.forEach((schemaTemplateFile) => {
-    templateJsonSchemaWithFiles(
-      path.join(taskSchemasDirectory, schemaTemplateFile),
-      path.join(taskSchemasDirectory, schemaTemplateFile.replace('.template', ''))
+async function runFilesCommand(
+  taskDirectory: string,
+  rawOptions: string[],
+  replacements?: any
+): Promise<void> {
+  const taskSchemasDirectory = path.join(taskDirectory, 'schemas');
+  const options = parseOptions(taskDirectory, rawOptions);
+  const schemaFiles = await fs.readdir(taskSchemasDirectory);
+  schemaFiles.filter((filename) => filename.endsWith('.template'))
+    .forEach(
+      (schemaTemplateFile) => {
+        const inputFile = path.join(taskSchemasDirectory, schemaTemplateFile);
+        const outputFile = path.join(
+          taskSchemasDirectory,
+          schemaTemplateFile.replace('.template', '')
+        );
+        templateJsonSchemaWithFiles(inputFile, outputFile, replacements);
+      }
     );
-  });
+  if (options.tsFile) {
+    const outputFile = await options.tsFile;
+    await generateTypes(taskSchemasDirectory, outputFile);
+  }
 }
+
+async function main() {
+  const taskDirectory = process.argv[2];
+  const command = process.argv[3];
+  const options = process.argv.slice(4);
+
+  switch (command) {
+    case 'files':
+      await runFilesCommand(taskDirectory, options);
+      return;
+    case 'queueGranulesFiles':
```

Comment:
What't the benefit to have this?

Reply:
I personally think the benefit is that it generates the correct schema for queue granules. What do you think would be a good alternative? 🙂

Reply:
The current (master) queue-granules schemas are fine, we don't need to use any templates.